runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -835,67 +837,153 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
-         }
-      else if (canTransformUnflattenedVTArrayElementLoadStore
-               || (arrayConstraint != NULL && isCompTypeVT == TR_no)
-               || (isStoreFlattenableArrayElement && isStoreValueVT == TR_no))
-         {
-         // If the array is known to be of a value type that is not flattened,
-         // or the array's component type is definitely not a value type, or if the value
-         // being assigned in an array store operation is definitely not a value type, add
-         // a delayed transformation to replace the helper call with inline code to
-         // perform the array element access.
+         // Determine whether the arraylength is known and whether a BNDCHK is required
          //
-         flags8_t flagsForTransform(isLoadFlattenableArrayElement ? ValueTypesHelperCallTransform::IsArrayLoad
-                                                                  : ValueTypesHelperCallTransform::IsArrayStore);
-         flagsForTransform.set(ValueTypesHelperCallTransform::InsertDebugCounter);
+         bool isBoundCheckRequired = true;
+         int32_t arrayLength = -1;
 
-         if (isStoreFlattenableArrayElement && !owningMethodDoesNotContainStoreChecks(this, node))
+         if (owningMethodDoesNotContainBoundChecks(this, node))
             {
-            // If storing to an array whose component type is or might be a value type
-            // and the value that's being assigned is or might be null, both a run-time
-            // NULLCHK of the value is required (guarded by a check of whether the
-            // component type is a value type) and an ArrayStoreCHK are required;
-            // otherwise, only the ArrayStoreCHK is required.
-            //
-            if ((isCompTypeVT != TR_no) && (storeValueConstraint == NULL || !storeValueConstraint->isNonNullObject()))
+            isBoundCheckRequired = false;
+            }
+         else if (arrayConstraint != NULL)
+            {
+            int32_t lowerBoundLimit;
+            int32_t upperBoundLimit;
+            int32_t elementSize;
+            bool isKnownObj;
+
+            getArrayLengthLimits(arrayConstraint, lowerBoundLimit, upperBoundLimit, elementSize, isKnownObj);
+
+            if (lowerBoundLimit == upperBoundLimit)
                {
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresNullValueCheck);
+               arrayLength = lowerBoundLimit;
                }
-            else
+
+            TR::VPConstraint *indexConstraint = getConstraint(indexNode, arrayRefGlobal);
+            if (indexConstraint != NULL)
                {
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
+               if (indexConstraint->getLowInt() >= 0 && indexConstraint->getHighInt() < lowerBoundLimit)
```

Comment:
Could a comment be added to explain why this condition determines bound check is not required?

Reply:
Fixed in commit 3e3941b73798c35167fcb8da13ea729159c3914d